File: Backend/app/schemas/client/triage.py
from pydantic import Field
from typing import Optional, List
from ..base import BaseSchema


class TriageChatRequest(BaseSchema):
    """分诊对话请求"""
    message: str = Field(..., min_length=1, max_length=500, description="患者症状描述")
    session_id: Optional[str] = Field(None, description="会话ID，首次对话可为空，后续轮次必填")


class DepartmentRecommendation(BaseSchema):
    """推荐科室"""
    department_id: int
    department_name: str
    confidence: str = Field(..., description="推荐置信度：high/medium/low")


class TriageChatResponse(BaseSchema):
    """分诊对话响应"""
    session_id: str = Field(..., description="会话ID，客户端需在后续请求中携带")
    message: str = Field(..., description="AI回复内容")
    recommendations: List[DepartmentRecommendation] = Field(default_factory=list, description="推荐科室列表")
    turn_count: int = Field(..., description="当前对话轮次（从1开始）")

File: medical-client/src/api/triage.ts
import request from '../utils/request'
import { clearToken, getToken } from '../utils/request'

export interface DepartmentRecommendation {
  department_id: number
  department_name: string
  confidence: 'high' | 'medium' | 'low'
}

export interface TriageChatRequest {
  message: string
  session_id?: string
}

export interface TriageChatResponse {
  session_id: string
  message: string
  recommendations: DepartmentRecommendation[]
  turn_count: number
}

export function triageChat(data: TriageChatRequest): Promise<TriageChatResponse> {
  return request.post('/triage/chat', data)
}

export type TriageStreamEvent =
  | { type: 'start'; session_id: string; turn_count: number }
  | { type: 'delta'; content: string }
  | { type: 'replace'; content: string }
  | { type: 'done'; session_id: string; turn_count: number; recommendations: DepartmentRecommendation[] }
  | { type: 'error'; message: string }

export async function triageChatStream(
  data: TriageChatRequest,
  onEvent: (event: TriageStreamEvent) => void,
  signal?: AbortSignal,
): Promise<void> {
  const token = getToken()
  const response = await fetch('/api/v1/triage/chat/stream', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(data),
    signal,
  })

  if (!response.ok) {
    let message = '智能分诊请求失败'
    try {
      const body = await response.json()
      message = body.message || body.detail || message
    } catch {
      // 非 JSON 错误响应沿用通用文案
    }
    if (response.status === 401 || response.status === 403) {
      clearToken()
      window.location.href = '/login'
    }
    throw new Error(message)
  }

  if (!response.body) throw new Error('浏览器不支持流式响应')

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  while (true) {
    const { value, done } = await reader.read()
    buffer += decoder.decode(value, { stream: !done })
    const lines = buffer.split('\n')
    buffer = lines.pop() || ''

    for (const line of lines) {
      if (line.trim()) onEvent(JSON.parse(line) as TriageStreamEvent)
    }
    if (done) break
  }

  if (buffer.trim()) onEvent(JSON.parse(buffer) as TriageStreamEvent)
}
